src/clp_ffi_js/ir/StreamReader.hpp

```diff
@@ -100,6 +115,108 @@ class StreamReader {
     [[nodiscard]] virtual auto decode_range(size_t begin_idx, size_t end_idx, bool use_filter) const
             -> DecodedResultsTsType = 0;
 
+    /**
+     * A templated function that implements `filter_log_events` for both
+     * `UnstructuredIrStreamReader` and `StructuredIrStreamReader`.
+     *
+     * @param log_level_filter
+     * @param log_events Derived class's log events.
+     * @param[out] filtered_log_event_map A reference to derived class's `FilteredLogEventsMap`
+     * that stores filtered result.
+     */
+    template <typename LogEvents>
+    static auto generic_filter_log_events(
+            FilteredLogEventsMap& filtered_log_event_map,
+            LogLevelFilterTsType const& log_level_filter,
+            LogEvents const& log_events
+    ) -> void {
+        if (log_level_filter.isNull()) {
+            filtered_log_event_map.reset();
+            return;
+        }
+
+        filtered_log_event_map.emplace();
+        auto filter_levels
+                = emscripten::vecFromJSArray<std::underlying_type_t<LogLevel>>(log_level_filter);
+
+        for (size_t log_event_idx = 0; log_event_idx < log_events.size(); ++log_event_idx) {
+            auto const& log_event = log_events[log_event_idx];
+            if (std::ranges::find(
+                        filter_levels,
+                        clp::enum_to_underlying_type(log_event.get_log_level())
+                )
+                != filter_levels.end())
+            {
+                filtered_log_event_map->emplace_back(log_event_idx);
+            }
+        }
+    }
+
+    /**
+     * A templated function that implements `decode_range` for both
+     * `UnstructuredIrStreamReader` and `StructuredIrStreamReader`.
+     *
+     * @param begin_idx
+     * @param end_idx
+     * @param filtered_log_event_map Derived class's `FilteredLogEventsMap`.
+     * @param log_events Derived class's log events.
+     * @param use_filter
+     * @param log_event_to_string Lambda function which retrieves a string from a single log event.
+     * The derived class must implement lambda since `Unstructured` and `Structured`
+     * log events have different interfaces.
+     * @return
+     */
+    template <typename LogEvents, typename ToStringFunc>
+    static auto generic_decode_range(
+            size_t begin_idx,
+            size_t end_idx,
+            FilteredLogEventsMap const& filtered_log_event_map,
+            LogEvents const& log_events,
```

Comment:
```suggestion
    template <typename LogEvent, typename ToStringFunc>
    requires requires(ToStringFunc func, LogEvent const& log_event) {
        {
            func(log_event)
        } -> std::convertible_to<std::string>;
    }
    static auto generic_decode_range(
            size_t begin_idx,
            size_t end_idx,
            FilteredLogEventsMap const& filtered_log_event_map,
            std::vector<LogEventWithFilterData<LogEvent>> const& log_events,
```

This constrains the types more explicitly.